app/services/preserver.rb

```diff
@@ -0,0 +1,125 @@
+    @change_set = change_set
+    @storage_adapter = storage_adapter || default_storage_adapter
+    @change_set_persister = change_set_persister
+  end
+
+  def preserve!
+    preserve_binary_content
+    # Don't preserve children unless this is the first time it's being
+    # preserved. After that point any updates to the children will trigger them
+    # to preserve themselves, because their parent is set up to be.
+    if preservation_object.persisted?
+      preserve_metadata
+    else
+      preserve_metadata && preserve_children
+    end
+  end
+
+  def preserve_binary_content
+    binary_nodes.each do |binary_node|
+      next if binary_node.file_identifiers.blank?
+      next if preservation_object.binary_nodes.find { |x| x.preservation_copy_of_id == binary_node.id }
+      file_metadata = file_metadata_from_binary_node(binary_node)
+      uploaded_file = storage_adapter.upload(
+        file: File.open(Valkyrie::StorageAdapter.find_by(id: binary_node.file_identifiers.first).disk_path),
+        original_filename: file_metadata.label.first,
+        resource: resource
+      )
+      file_metadata.file_identifiers = uploaded_file.id
+      preservation_object.binary_nodes ||= []
+      preservation_object.binary_nodes += [file_metadata]
+    end
+  end
+
+  def binary_nodes
+    [:original_files, :intermediate_files, :preservation_files].flat_map do |node_type|
+      Array(resource.try(node_type))
+    end
+  end
+
+  def file_metadata_from_binary_node(node)
+    label, splitter, extension = node.label.first.to_s.rpartition(".")
+    label = "#{label}-#{node.id}#{splitter}#{extension}"
+    FileMetadata.new(
+      label: label,
+      use: Valkyrie::Vocab::PCDMUse.PreservationCopy,
+      mime_type: node.mime_type,
+      checksum: node.checksum,
+      preservation_copy_of_id: node.id
+    )
+  end
+
+  def preservation_object
+    @preservation_object ||=
+      begin
+        Wayfinder.for(resource).try(:preservation_object) || PreservationObject.new(preserved_object_id: resource.id)
+      end
+  end
+
+  def preserve_children
+    return unless resource.try(:member_ids).present?
+    PreserveChildrenJob.perform_later(id: resource.id.to_s)
+  end
+
+  def preserve_metadata
+    metadata_node = preservation_object.metadata_node || build_metadata_node
+    uploaded_file = storage_adapter.upload(file: temp_metadata_file.io, original_filename: metadata_node.label.first, resource: resource)
+    metadata_node.file_identifiers = uploaded_file.id
+    preservation_object.metadata_node = metadata_node
+    change_set_persister.metadata_adapter.persister.save(resource: preservation_object)
+  end
+
+  def metadata_checksum
+    @metadata_checksum ||=
+      begin
+        io = StringIO.new(preservation_metadata.to_json)
+        MultiChecksum.for(Valkyrie::StorageAdapter::File.new(io: io, id: "tmp"))
+      end
+  end
+
+  # Don't preserve the PreservedMetadata FileMetadataNode, because it's
```

Comment:
I feel like this comment might be more helpful somewhere else. I can't really figure out how it's relevant here. Or maybe it needs to be rewritten? I think it may help to have the associated ADR revised / nailed down; this could maybe point there?

Reply:
Oh, I think this is just an old comment. It's actually pretty irrelevant now.